### PR TITLE
Aggregate signatures: check completeness for signatures stored locally

### PR DESCRIPTION
**Pull request by rbauduin**

### CODE REVIEW

File: aggregate_signature/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,743 @@
...
+                    .map_err(|e| AggregateSignatureError::Signature(e.to_string()))?;
+
+                signatures.insert(pubkey, signature);
+            }
+        }
+
+        Ok(Self {
+            signatures,
+            origin: path.to_string_lossy().to_string(),
+        })
```
[the remark this answers is not in the document]
Comment (rbauduin) [reply aggregate_signature/src/lib.rs]: I don't think there's  a shadowing bug, because:

* the test reported as probably failing passes
* the shadowing variable does so in a scope where the shadowed variable is not used. When the `path` variable is used to set the `origin', there is no shadowing as it is done outside of the scope of the shadowing variable (defined in the `for...in` loop).

However, to avoid confusion, I have renamed both variable with more explicit names.

File: aggregate_signature/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,743 @@
...
+        Self::check_groups(&signers_config.master_keys, &self.signatures)
+    }
+
+    pub fn is_admin_complete(&self, signers_config: &SignersConfig<P>) -> bool {
+        // Check admin_keys groups if present
+        if let Some(admin_keys) = &signers_config.admin_keys {
+            Self::check_groups(admin_keys, &self.signatures)
+        } else {
+            false
+        }
```
[the remark this answers is not in the document]
Comment (rbauduin) [reply aggregate_signature/src/lib.rs]: The suggested refactor is [discouraged by clippy](https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_map_or).
Here's the better one: 
```
        signers_config
            .admin_keys
            .as_ref()
            .is_some_and(|keys| Self::check_groups(keys, &self.signatures))
```
I also added some tests to ensure empty admin_keys are not considered as complete.

File: aggregate_signature/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,748 @@
...
+        Self::check_groups(&signers_config.master_keys, &self.signatures)
+    }
+
+    pub fn is_admin_complete(&self, signers_config: &SignersConfig<P>) -> bool {
+        // Check admin_keys groups if present
+        signers_config
+            .admin_keys
+            .as_ref()
+            .is_some_and(|keys| Self::check_groups(keys, &self.signatures))
+    }
```
[the remark this answers is not in the document]
Comment (rbauduin) [reply aggregate_signature/src/lib.rs:90]: I had added a test to cover this case but made a mistake, testing with `None` and not `Some([])`...
I put the check for empty groups in `check_group` instead, and empty groups never validate.

File: aggregate_signature/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,748 @@
...
+
+        // Should be complete with mixed configuration
+        assert!(agg_sig.is_artifact_complete(&signers_config_mixed));
+        assert!(agg_sig.is_master_complete(&signers_config_mixed));
+        // Check an null admin group is not comsidered as complete
+        assert!(!agg_sig.is_admin_complete(&signers_config_mixed));
+        // Empty admin_keys array is never complete
+        let mut signers_config_empty_admin = signers_config_mixed.clone();
+        signers_config_empty_admin.admin_keys = None;
+        assert!(!agg_sig.is_admin_complete(&signers_config_mixed));
```
[the remark this answers is not in the document]
Comment (rbauduin) [reply aggregate_signature/src/lib.rs]: This is another part of my messup with the test of empty admin groups. It is also fixed.

File: aggregate_signature/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,759 @@
...
+        let keypair2 = AsfaloadKeyPair::new("password").unwrap();
+        let pubkey2 = keypair2.public_key();
+        let _seckey2 = keypair2.secret_key("password").unwrap();
+
+        // Create signature
+        let data = b"test data";
+        let signature = seckey.sign(data).unwrap();
+
+        // Write signature file (base64-encoded pubkey filename)
+        let pubkey_b64 = BASE64_STANDARD.encode(pubkey.to_base64());
```
[the remark this answers is not in the document]
Comment (rbauduin) [reply aggregate_signature/src/lib.rs]: Fixed, using base64url for filenames in `use base64url for filenames`.

File: aggregate_signature/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,759 @@
...
+        // Empty admin_keys array is never complete
+        let mut signers_config_empty_admin = signers_config_mixed.clone();
+        signers_config_empty_admin.admin_keys = Some([].to_vec());
+        assert!(!agg_sig.is_admin_complete(&signers_config_empty_admin));
+
+        assert_eq!(agg_sig.origin, "test_origin");
+    }
+
+    #[test]
+    fn test_check_groups_from_json_minimal() {
```
[the remark this answers is not in the document]
Comment (rbauduin) [reply aggregate_signature/src/lib.rs:385]: updated issue #3 with suggestion

File: aggregate_signature/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,15 @@
+[package]
+name = "aggregate_signature"
+version = "0.1.0"
+edition = "2024"
```
[the remark this answers is not in the document]
Comment (rbauduin) [reply aggregate_signature/Cargo.toml:4]: let's keep 2024 and change if we encounter problems

File: signatures/src/keys/minisign.rs [hunk 8]
```diff
@@ -13,7 +13,6 @@ use crate::keys::{
 };
 
 pub mod errs {
-    use base64::DecodeError;
```
[the remark this answers is not in the document]
Comment (rbauduin) [reply signatures/src/keys/minisign.rs:16]: This is ok because the use of DecodeError is fully qualified.

File: aggregate_signature/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,756 @@
...
+      }
+    ]
+    "#,
+                signatures_1_2_3_4.clone(),
+                complete,
+            ),
+            //------------------------------------------------------------
+            // 2-of-1: this will never be complete, even if we have the signature from the
+            // signer and an additional one.
+            // FIXME: this json should raise an error when parsed to build our rust object.
```
[the remark this answers is not in the document]
Comment (rbauduin) [reply aggregate_signature/src/lib.rs:579]: Added code suggestion to issue #3
